ASSERT on Linux: x64 tests/free shadow.c:261: false (fail to create shadow memory mapping)

This happened once in a suite run, and I can't reproduce it:

```
drheapstat-dbg-32: all 30 tests passed
drheapstat-rel-32: all 30 tests passed
drmemory-dbg-32: all 67 tests passed
drmemory-dbg-64: 13 tests passed, **** 1 tests failed: ****
        free => ~~Dr.M~~ ASSERT FAILURE (thread 14198): /work/drmemory/git/src/drmemory/shadow.c:261: false (fail to create shadow memory mapping)

                CMake Error at runtest.cmake:311 (message):;  ~~Dr.M~~ ASSERT FAILURE (thread 14198):
                  /work/drmemory/git/src/drmemory/shadow.c:261: false (fail to create shadow
                  memory mapping)
drmemory-rel-32: all 67 tests passed
drmemory-rel-64: all 14 tests passed
final package: build successful; no tests for this build
```

--- a/tests/framework/CMakeLists.txt
+++ b/tests/framework/CMakeLists.txt
@@ -1,5 +1,5 @@
 # **********************************************************
-# Copyright (c) 2012-2020 Google, Inc.  All rights reserved.
+# Copyright (c) 2012-2021 Google, Inc.  All rights reserved.
 # **********************************************************
 
 # Dr. Memory: the memory debugger
@@ -56,7 +56,8 @@ endfunction(add_drmf_test_app)
 # We only expect a few tests, so we simplify things by having a simple
 # regex output for whether they passed.
 # ext_list should be a list of extensions minus the drmf_ prefix.
-function(add_drmf_test test_name app_name src_client ext_list client_options pass_regex)
+function(add_drmf_test test_name app_name src_client ext_list dr_options
+    client_options pass_regex)
   set(client_name ${test_name}.client)
   add_library(${client_name} SHARED ${src_client})
   _DR_append_property_list(TARGET ${client_name} COMPILE_DEFINITIONS "${arch_defs}")
@@ -95,7 +96,7 @@ function(add_drmf_test test_name app_name src_client ext_list client_options pas
     add_test(${test_name}
       ${drrun_path} ${drrun_extra}
       -client ${client_path} 0 "${client_options}"
-      -msgbox_mask 0 -stderr_mask 0xc
+      -msgbox_mask 0 -stderr_mask 0xc ${dr_options}
       "--" ${app_path})
     set_tests_properties(${test_name} PROPERTIES PASS_REGULAR_EXPRESSION "${pass_regex}")
   endif ()
@@ -110,12 +111,12 @@ set(symcache_dir "${PROJECT_BINARY_DIR}/logs/symcache")
 
 add_drmf_test_app(drsyscall_app drsyscall_app.c)
 add_drmf_test(drsyscall_test drsyscall_app drsyscall_client.c
-  drsyscall "${symcache_dir}" "done\nTEST PASSED")
+  drsyscall "" "${symcache_dir}" "done\nTEST PASSED")
 if (NOT ANDROID) # XXX i#1860: Android tests not enabled yet.
   set_property(TEST drsyscall_test APPEND PROPERTY DEPENDS hello)
 endif ()
 add_drmf_test(strace_test drsyscall_app strace_client.c
-  drsyscall "${symcache_dir}" "done\n.*TEST PASSED")
+  drsyscall "" "${symcache_dir}" "done\n.*TEST PASSED")
 if (NOT ANDROID) # XXX i#1860: Android tests not enabled yet.
   set_property(TEST strace_test APPEND PROPERTY DEPENDS hello)
 endif ()
@@ -123,13 +124,13 @@ endif ()
 # drfuzz tests
 add_drmf_test_app(drfuzz_app_empty drfuzz_app_empty.c)
 add_drmf_test(drfuzz_test_empty drfuzz_app_empty drfuzz_client_empty.c
-  drfuzz "" "done\nTEST PASSED")
+  drfuzz "" "" "done\nTEST PASSED")
 add_drmf_test(drfuzz_test_mutator drfuzz_app_empty drfuzz_client_mutator.c
-  drfuzz "" "TEST PASSED\n.*done")
+  drfuzz "" "" "TEST PASSED\n.*done")
 
 add_drmf_test_app(drfuzz_app_repeat drfuzz_app_repeat.c)
 add_drmf_test(drfuzz_test_repeat drfuzz_app_repeat drfuzz_client_repeat.c
-  drfuzz "" "hello 1\nhello 2\nhello 3\nhello 4\nhello 5\ndone\nTEST PASSED")
+  drfuzz "" "" "hello 1\nhello 2\nhello 3\nhello 4\nhello 5\ndone\nTEST PASSED")
 
 add_drmf_test_app(drfuzz_app_segfault drfuzz_app_segfault.c)
 
@@ -138,15 +139,15 @@ set(segfault_regex_2 "Crash originated in target.*with 1 args.*")
 set(segfault_regex_3 "TEST PASSED")
 
 add_drmf_test(drfuzz_test_segfault drfuzz_app_segfault drfuzz_client_segfault.c
-  drfuzz "-crash" "${segfault_regex_1}${segfault_regex_2}${segfault_regex_3}")
+  drfuzz "" "-crash" "${segfault_regex_1}${segfault_regex_2}${segfault_regex_3}")
 use_DynamoRIO_extension(drfuzz_test_segfault.client drsyms)
 
 add_drmf_test(drfuzz_test_app_abort drfuzz_app_segfault drfuzz_client_segfault.c
-  drfuzz "-abort" "${segfault_regex_2}${segfault_regex_3}")
+  drfuzz "" "-abort" "${segfault_regex_2}${segfault_regex_3}")
 use_DynamoRIO_extension(drfuzz_test_app_abort.client drsyms)
 
 add_drmf_test(drfuzz_test_no_crash drfuzz_app_segfault drfuzz_client_segfault.c
-  drfuzz "" ".*TEST PASSED")
+  drfuzz "" "" ".*TEST PASSED")
 use_DynamoRIO_extension(drfuzz_test_no_crash.client drsyms)
 
 # XXX i#1734: add multi-threaded test, checking for aborted fuzz targets on all threads
@@ -163,18 +164,22 @@ if (WIN32)
 endif ()
 target_include_directories(umbra_app PRIVATE ${CMAKE_CURRENT_SOURCE_DIR})
 
-add_drmf_test(umbra_test_empty      umbra_app umbra_client_empty.c
-  umbra "" ".*TEST PASSED")
+add_drmf_test(umbra_test_empty umbra_app umbra_client_empty.c
+  umbra "" "" ".*TEST PASSED")
+
+# Test i#1712 with a region overlapping 0x7e00'-0x7f00'.
+add_drmf_test(umbra_test_overlap umbra_app umbra_client_empty.c
+  umbra "-vm_base;0x7efff0000000;-no_vm_base_near_app" "" ".*TEST PASSED")
 
 add_drmf_test(umbra_test_shadow_mem umbra_app umbra_client_shadow_mem.c
-  umbra "" ".*TEST PASSED")
+  umbra "" "" ".*TEST PASSED")
 use_DynamoRIO_extension(umbra_test_shadow_mem.client drreg)
 use_DynamoRIO_extension(umbra_test_shadow_mem.client drutil)
 target_include_directories(umbra_test_shadow_mem.client PRIVATE
   ${CMAKE_CURRENT_SOURCE_DIR})
 
 add_drmf_test(umbra_test_insert_app_to_shadow umbra_app
-  umbra_client_insert_app_to_shadow.c umbra "" ".*TEST PASSED")
+  umbra_client_insert_app_to_shadow.c umbra "" "" ".*TEST PASSED")
 use_DynamoRIO_extension(umbra_test_insert_app_to_shadow.client drreg)
 use_DynamoRIO_extension(umbra_test_insert_app_to_shadow.client drutil)
 target_include_directories(umbra_test_insert_app_to_shadow.client PRIVATE
@@ -183,7 +188,7 @@ target_include_directories(umbra_test_insert_app_to_shadow.client PRIVATE
 # Faulty redzones are only available on 32-bit.
 if (NOT X64)
   add_drmf_test(umbra_client_faulty_redzone umbra_app
-    umbra_client_faulty_redzone.c umbra "" ".*TEST PASSED")
+    umbra_client_faulty_redzone.c umbra "" "" ".*TEST PASSED")
   use_DynamoRIO_extension(umbra_client_faulty_redzone.client drreg)
   use_DynamoRIO_extension(umbra_client_faulty_redzone.client drutil)
   target_include_directories(umbra_client_faulty_redzone.client PRIVATE
@@ -196,11 +201,11 @@ if (NOT X64)
 endif()
 
 add_drmf_test(umbra_test_consistency umbra_app
-  umbra_client_consistency.c umbra "" ".*TEST PASSED")
+  umbra_client_consistency.c umbra "" "" ".*TEST PASSED")
 use_DynamoRIO_extension(umbra_test_consistency.client drreg)
 use_DynamoRIO_extension(umbra_test_consistency.client drutil)
 target_include_directories(umbra_test_consistency.client PRIVATE
   ${CMAKE_CURRENT_SOURCE_DIR})
 
 add_drmf_test(umbra_test_allscales umbra_app umbra_client_allscales.c
-  umbra "" ".*TEST PASSED")
+  umbra "" "" ".*TEST PASSED")

--- a/tests/framework/CMakeLists.txt
+++ b/tests/framework/CMakeLists.txt
@@ -167,9 +167,11 @@ target_include_directories(umbra_app PRIVATE ${CMAKE_CURRENT_SOURCE_DIR})
 add_drmf_test(umbra_test_empty umbra_app umbra_client_empty.c
   umbra "" "" ".*TEST PASSED")
 
-# Test i#1712 with a region overlapping 0x7e00'-0x7f00'.
-add_drmf_test(umbra_test_overlap umbra_app umbra_client_empty.c
-  umbra "-vm_base;0x7efff0000000;-no_vm_base_near_app" "" ".*TEST PASSED")
+if (X64)
+  # Test i#1712 with a region overlapping 0x7e00'-0x7f00'.
+  add_drmf_test(umbra_test_overlap umbra_app umbra_client_empty.c
+    umbra "-vm_base;0x7efff0000000;-no_vm_base_near_app" "" ".*TEST PASSED")
+endif ()
 
 add_drmf_test(umbra_test_shadow_mem umbra_app umbra_client_shadow_mem.c
   umbra "" "" ".*TEST PASSED")

--- a/umbra/umbra_64.c
+++ b/umbra/umbra_64.c
@@ -1,5 +1,5 @@
 /* **********************************************************
- * Copyright (c) 2013-2020 Google, Inc.  All rights reserved.
+ * Copyright (c) 2013-2021 Google, Inc.  All rights reserved.
  * **********************************************************/
 
 /* Dr. Memory: the memory debugger
@@ -492,6 +492,15 @@ umbra_add_shadow_segment(umbra_map_t *map, app_segment_t *seg)
         /* new app-seg vs other app-seg's shadow and reserve */
         base = seg->app_base;
         end  = seg->app_end;
+        if (seg != &app_segments[i] &&
+            segment_overlap(base, end, app_segments[i].app_base,
+                            app_segments[i].app_end)) {
+            ELOG(1, "ERROR: new app segment ["PFX", "PFX")"
+                 " conflicts with app seg ["PFX", "PFX")\n",
+                 seg->app_base, seg->app_end,
+                 app_segments[i].app_base, app_segments[i].app_end);
+            return false;
+        }
         for (map_idx = 0; map_idx < MAX_NUM_MAPS; map_idx++) {
             if (app_segments[i].map[map_idx] == NULL)
                 continue;
@@ -502,7 +511,7 @@ umbra_add_shadow_segment(umbra_map_t *map, app_segment_t *seg)
                                 app_segments[i].reserve_base[map_idx],
                                 app_segments[i].reserve_end[map_idx])) {
                 ELOG(1, "ERROR: new app segment ["PFX", "PFX")"
-                     " conflicts with app seg ["PFX", "PFX") or its "
+                     " conflicts with app seg ["PFX", "PFX")'s "
                      "shadow ["PFX", "PFX") or reserve ["PFX", "PFX")\n",
                      seg->app_base, seg->app_end,
                      app_segments[i].app_base, app_segments[i].app_end,
@@ -579,6 +588,12 @@ umbra_add_app_segment(app_pc base, size_t size, umbra_map_t *map)
     uint i;
 
     LOG(UMBRA_VERBOSE, "add new app segment for ["PFX", "PFX")\n", base, base + size);
+    app_pc seg_base = (app_pc)segment_base(num_seg_bits, base);
+    /* We do support a memory range spanning multiple segments, since it
+     * happens in practice with Control Flow Guard (i#2184).
+     */
+    app_pc seg_end =
+        (app_pc)ALIGN_FORWARD(base + size, segment_size(num_seg_bits));
     for (i = 0; i < MAX_NUM_APP_SEGMENTS; i++) {
         if (app_segments[i].app_used) {
             if (base >= app_segments[i].app_base &&
@@ -586,19 +601,49 @@ umbra_add_app_segment(app_pc base, size_t size, umbra_map_t *map)
                 return true;
         } else {
             if (app_segments[i].app_end != NULL) {
-                /* a pre-defined app segment */
+                /* Entirely inside a pre-defined app segment? */
                 if (base >= app_segments[i].app_base &&
                     base + size <= app_segments[i].app_end) {
                     app_segments[i].app_used = true;
                     return true;
                 }
+                /* Overlaps a pre-defined app segment on the left? */
+                if (base + size > app_segments[i].app_base &&
+                    base + size < app_segments[i].app_end) {
+                    LOG(1, "adjusting pre-defined app segment [%p, %p) to [%p, %p)"
+                        " to incorporate [%p-%p)\n",
+                        app_segments[i].app_base, app_segments[i].app_end,
+                        seg_base, app_segments[i].app_end, base, base+size);
+                    for (int j = 0; j < MAX_NUM_APP_SEGMENTS; j++) {
+                        if (app_segments[j].app_used &&
+                            app_segments[j].app_base == seg_base &&
+                            app_segments[j].app_end == app_segments[i].app_base) {
+                            LOG(1, "invalidating now-merged app segment [%p, %p)\n",
+                                app_segments[j].app_base, app_segments[j].app_end);
+                            app_segments[j].app_used = false;
+                            app_segments[j].app_base = NULL;
+                            app_segments[j].app_end = NULL;
+                        }
+                    }
+                    app_segments[i].app_base = seg_base;
+                    app_segments[i].app_used = true;
+                    return true;
+                }
+                /* Overlaps a pre-defined app segment on the right? */
+                if (base > app_segments[i].app_base &&
+                    base < app_segments[i].app_end) {
+                    LOG(1, "adjusting pre-defined app segment [%p, %p) to [%p, %p)"
+                        " to incorporate [%p-%p)\n",
+                        app_segments[i].app_base, app_segments[i].app_end,
+                        app_segments[i].app_base, seg_end, base, base+size);
+                    /* No invalidation check, since we're walking left to right. */
+                    app_segments[i].app_end = seg_end;
+                    app_segments[i].app_used = true;
+                    return true;
+                }
             } else {
-                app_segments[i].app_base = (app_pc)segment_base(num_seg_bits, base);
-                /* We do support a memory range spanning multiple segments, since it
-                 * happens in practice with Control Flow Guard (i#2184).
-                 */
-                app_segments[i].app_end =
-                    (app_pc)ALIGN_FORWARD(base + size, segment_size(num_seg_bits));
+                app_segments[i].app_base = seg_base;
+                app_segments[i].app_end = seg_end;
                 LOG(1, "adding app segment ["PFX", "PFX")\n", app_segments[i].app_base,
                     app_segments[i].app_end);
                 /* Adding a not pre-defined segment.

